Complete Item Endpoints

lao9 opened the pull request:
Did a couple of things here - got item find working (except for dates...) and wanted to run spec harness - still a bunch of failing tests there

josh-works committed:
Transactions, customer migrations

I'm not 100% convinced I defined the datatypes correctly for our
customers, transactions, and merchants models.

That said, this was partially intentional, as I wanted practice making
regular and minor changes to our schemas.

So, since we don't have any relationships between the models, there's a
column for some foreign keys, but no actual data.

I look forward to resolving this.

Our rake tasks feel clunky and wrong, but it'll do for now.

lao9 committed:
add item model tests

closes issue #6
created factory
added tables to database
added gems for shoulda-matchers and faker

lao9 committed:
write tests and funcitonality for invoice index

create new factory for invoices
update merchant factory to have unique-ish titles
figure out how to create a json relationship for next iteration

lao9 committed:
write tests and functionality for invoice show

josh-works committed:
add factory for customers

lao9 committed:
update invoice item relationships

lao9 committed:
write tests and functionality for invoice item api

josh-works committed:
add validations, associations

customers, merchants now have tests (that pass) for:
validations of all fields
associations for many-to-many relationships

lao9 committed:
added directories to contain many controllers

updated view file structure
updated controller file structure
affects items, invoices, and invoice items

josh-works committed:
add relationships to transactions

added validations and relationships to transactions class
updated tests to use factories

josh-works committed:
fix merge conflict

josh-works committed:
Merge remote-tracking branch 'origin/lo-invoice-item-import' into 1_data_import

lao9 committed:
update items show to pass a few spec harness tests

got priliminary find completed

case-eee commented on Gemfile:
```diff
@@ -28,6 +28,14 @@ gem 'puma', '~> 3.0'
+  gem 'database_cleaner'
+  gem 'shoulda-matchers'
+  gem 'faker', :git => 'git://github.com/stympy/faker.git', :branch => 'master'
+  gem 'harry_potter_faker'
```
🎉

case-eee commented on app/controllers/api/v1/customers_controller.rb:
```diff
@@ -0,0 +1,11 @@
+class Api::V1::CustomersController < ApplicationController
+
+  def index
+    render json: Customer.all, only: [:id, :first_name, :last_name]
```
i'd suggest using a serializer here (i'm assuming y'all just haven't gotten to it yet so no worries!)

case-eee commented on app/controllers/api/v1/items/items_find_controller.rb:
```diff
@@ -0,0 +1,18 @@
+
+  private
+
+  def search_params
```
👍

case-eee commented on app/models/invoice.rb:
```diff
@@ -0,0 +1,10 @@
+  belongs_to :merchant
+  has_many :transactions
+  has_many :invoice_items
+  has_many :items, through: :invoice_items
```
woooo AR associations 🎉

case-eee commented on app/models/item.rb:
```diff
@@ -0,0 +1,12 @@
+  has_many :invoice_items
+  has_many :invoices, through: :invoice_items
+
+  def format_unit_price
```
👍

case-eee commented on app/views/api/v1/items/items/index.json.jbuilder:
```diff
@@ -0,0 +1,3 @@
+json.(@items) do |item|
```
woooo jbuilder. good work 👍

case-eee commented on lib/tasks/load_data.rake:
```diff
@@ -0,0 +1,44 @@
+
+    puts "loading customers"
+    CSV.foreach('db/data/customers.csv', :headers => true, :encoding => 'ISO-8859-1:UTF-8') do |row|
+      Customer.create([
```
when seeding data or in your tests, it's a good idea to use the bang methods - so using `create!` instead of `create`  - this will cause ActiveRecord to fail loudly and let you know if an `Item` is not created. as it is now, if a specific row doesn't get translated to an `Item`, you won't actually know about it!

case-eee commented on spec/models/invoice_spec.rb:
```diff
@@ -0,0 +1,21 @@
+  describe "is valid with merchant, customer, and status" do
+    it { should validate_presence_of(:status) }
+  end
+  describe "relationships" do
```
👍

case-eee commented on spec/models/invoice_spec.rb:
```diff
@@ -0,0 +1,21 @@
+require 'rails_helper'
+
+RSpec.describe Invoice, type: :model do
+  describe "is valid with merchant, customer, and status" do
```
we should probably validate that it has a merchant and customer along with a status!

case-eee commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,20 @@
+require 'rails_helper'
+
+RSpec.describe Item, type: :model do
```
be sure you test your `format_unit_price` method as well!

lao9 replied:
gooood catch @case-eee !

case-eee replied:
perks of having an extra set of eyes on code 👍

case-eee reviewed (commented):
@lao9 @josh-works left y'all some comments. good work so far 👍 keep it up!

lao9 committed:
finish find and find all functionality

still need to finish tests for find all but all tests are passing via spec harness

lao9 committed:
finish items find all tests

lao9 committed:
finish random tests and functionality for items

case-eee commented on app/controllers/api/v1/items/items_random_controller.rb:
```diff
@@ -0,0 +1,8 @@
+
+  def show
+    id = Item.pluck(:id).sample
+    @item = Item.find(id)
```
This logic is fine, but does it belong in your controller? Where could we put this that's more appropriate? Perhaps a "random" method

lao9 committed:
refactored application record and controller

removed formatting from item to app record module
created before action callback to run and edit params

lao9 committed:
updated rake seed file

lao9 committed:
resolve merge conflicts

josh-works commented:
all looks great. We've added @case-eee's suggestions as waffle.io cards, and will address later.

case-eee commented:
glad you added cards for those things - that's perfect @josh-works 👍
